[dsja612] iP

dsja612 opened the pull request:
# Duke
> "Quotations on the Internet are often misattributed." - Duke

# Yet another school project??
Duke is a task manager written in Java, in an object-oriented manner. This is all so you can live your life in a task-oriented manner 🤓

# "Pfft, why would I use Duke instead of Google Calendar??" - 🤓
- Text-based, specifically for tech 🤓s.
- Easy to learn
- As fast as you can type

And it is **FREE!**

# Features
- [x] CLI based (for now)
- [x] Saves your tasks even after you exit!
- [x] Everything is abstracted away just like OOP!

Download it [here!](https://github.com/dsja612/ip/releases/tag/A-Jar)

## If you want to learn Java & OOP:
Feel free to check out how Duke is organized. Duke starts running from this code block in `Duke.java`:
```ruby
    public static void main(String[] args) throws IOException {
        Duke.run();
    }
```

Commit:
Add Gradle support

dsja612 committed:
Add greeting, echo and exit

dsja612 committed:
Add Task.java, TaskList.java, adding and listing tasks

dsja612 committed:
Add marking of tasks as done or undone

dsja612 committed:
Add ToDo, Event, Deadline

dsja612 committed:
Add automated text UI testing

dsja612 committed:
Add error handling for incorrect user input

dsja612 committed:
Add error handling for mark and unmark

dsja612 committed:
Rename some variables to be more descriptive

dsja612 committed:
Add enum to represent type of commands

dsja612 committed:
Add deleting of tasks and more error handling

dsja612 committed:
Move print statements to Duke.java

dsja612 committed:
Fix UI bugs

dsja612 committed:
Add storage file

dsja612 committed:
Add date support for events and deadlines

dsja612 committed:
Add datetime support for events and deadlines

dsja612 committed:
Add local file storage of tasks

dsja612 committed:
Merge branch 'branch-Level-7'

dsja612 committed:
Merge branch 'branch-Level-8'

dsja612 committed:
Move class files into packages

dsja612 committed:
Add UI class to handle interactions with user

dsja612 committed:
Organize code to be more object oriented

dsja612 committed:
Merge branch 'add-gradle-support' of https://github.com/dsja612/ip

dsja612 committed:
Integrate Gradle into project

dsja612 committed:
Format UI text display

dsja612 committed:
Fix formatting for tasks with date

dsja612 committed:
Add J-Unit test

dsja612 committed:
Add build items to gitignore

dsja612 committed:
Add find command to search for tasks using keywords

dsja612 committed:
Add JavaDocs for all classes except for command subclasses

dsja612 committed:
Made code comply with Java coding standards

dsja612 committed:
Merge branch 'Level-9'

dsja612 committed:
Merge branch 'A-JavaDoc'

dsja612 committed:
Merge branch 'A-CodingStandard'

dsja612 committed:
Fix merge conflicts

dsja612 committed:
Change more code to adhere to coding standards

immanuelhume commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,50 @@
+    private static final TaskList tasks = new TaskList();
+    private static final Scanner sc = new Scanner(System.in);
+    private static final UI ui = new UI();
+    private static final Storage storage = new Storage(tasks);
```
Since these variables are static, perhaps it is best to make the constructor for the `Duke` class private, and make it a final class? This prevents multiple Duke instances from being created.

immanuelhume commented on src/main/java/commands/EventCommand.java:
```diff
@@ -0,0 +1,42 @@
+            this.endDate = LocalDateTime.parse(endDate, DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm"));
+            if (this.startDate.isAfter(this.endDate)) {
+                throw new DukeException("The starting date of this event is after its ending date!");
+            }
```
Maybe we can move this `if` block out of the `try-catch` block, to reduce the amount of nesting.

immanuelhume commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,38 @@
+    @Override
+    public void execute(UI ui, TaskList tasks, Storage storage) throws DukeException {
+        if (tasks.size() == 0) {
+            this.commandStatus = "You have no tasks in your list! \n";
```
Could consider adding an early `return` here, then we can drop the else block.

immanuelhume commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,109 @@
+ * Handles loading and saving of tasks to local storage.
+ */
+public class Storage {
+    private final Path HOME_DIRECTORY = Path.of(System.getProperty("user.dir") + "/data");
```
I guess this variable name might be misunderstood to mean `$HOME` on unix. Would `DUKE_DATA_DIR` be better?

immanuelhume commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,109 @@
+            case "T":
+                tasks.addTask(new ToDo(taskName));
+                break;
+            case "D":
+                String by = inputList[2];
+                LocalDateTime deadline = LocalDateTime.parse(by, DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm"));
+                tasks.addTask(new Deadline(taskName, deadline));
+                break;
```
Regarding variable declarations in `case` blocks, it might be better to wrap them inside their own scope. This prevents potential name clashes with other branches of the `switch` statement.

```suggestion
            case "D": {
                String by = inputList[2];
                LocalDateTime deadline = LocalDateTime.parse(by, DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm"));
                tasks.addTask(new Deadline(taskName, deadline));
                break;
            }
```

dsja612 replied:
Great observation, thanks! I wasn't sure of whether to add braces for every block as that would further increase the indentation of the code

resolved

immanuelhume commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+        StringBuilder result = new StringBuilder();
+        result.append("List of Tasks: \n");
+        for (int i = 0; i < taskList.size(); i++) {
+            result.append((i + 1));
+            result.append(". ");
+            result.append(taskList.get(i).toString());
+            result.append("\n");
```
Could consider chaining methods here so it's less verbose.

```suggestion
            result.append((i + 1))
                    .append(". ")
                    .append(taskList.get(i).toString())
                    .append("\n");
```

immanuelhume reviewed (commented):
Nice work bro :+1:

dsja612 committed:
Change date format in file parser to be consistent with format in local storage

dsja612 committed:
Update src/main/java/tasks/TaskList.java

Co-authored-by: Junyi <47441677+immanuelhume@users.noreply.github.com>

dsja612 committed:
Add braces to cases in switch statements

dsja612 committed:
Change access modifiers in Duke.java

dsja612 committed:
Rename variables in Storage to improve clarity

sumhungyee commented on src/main/java/views/UI.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    /**
+     * Displays a list of possible commands
+     * to the user
```
Might it be better to end with a fullstop?

sumhungyee commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,45 @@
+
+    /**
+     * Constructor for an Event task.
+     * @param name Name of the task
+     * @param startDate Starting date of the task
+     * @param endDate Ending date of the task
```
Might it be better to add a fullstop behind these lines?

sumhungyee commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,113 @@
+            }
+            case "D": {
+                String by = inputList[2];
+                LocalDateTime deadline = LocalDateTime.parse(by, DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm"));
```
Is this line a bit too long?

sumhungyee commented:
Great job dude 👍

sumhungyee reviewed (commented):
Overall, I find your code easy to read! For the most part, it looks like coding standards were upheld! Great Job!

sumhungyee commented on src/main/java/exceptions/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+
+    /**
+     * Constructor for DukeException that stores an error message.
+     * @param message Main body of error message
```
Perhaps this needs capitalising as well!

sumhungyee commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,113 @@
+
+    /**
+     * Loads tasks from a local file into a TaskList, creates one if one does not exist.
+     * @throws IOException If fail to load/create a storage file
```
Perhaps this needs capitalising as well!

sumhungyee commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,113 @@
+     * @throws DukeException If fail to create task from invalid input
+     */
+    public static void parseFile(String input, TaskList tasks) throws DukeException {
+        String[] inputList = input.split(",");
+        String taskType = inputList[0];
+        String taskName = inputList[1];
```
Your code here looks extremely readable! 👓

sumhungyee commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,113 @@
+        String taskType = inputList[0];
+        String taskName = inputList[1];
+
+        switch (taskType) {
+            case "T": {
+                tasks.addTask(new ToDo(taskName));
+                break;
+            }
+            case "D": {
+                String by = inputList[2];
+                LocalDateTime deadline = LocalDateTime.parse(by, DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm"));
+                tasks.addTask(new Deadline(taskName, deadline));
+                break;
+            }
```
The case statements are at a wrong indentation level! Perhaps you may need to shift them down one level!

dsja612 replied:
Good catch!

sumhungyee reviewed (commented):
Well done!  Your code looks fantastic!

dsja612 committed:
Change code to conform to CheckStyle specs

dsja612 committed:
Merge branch 'branch-A-CheckStyle'

dsja612 committed:
Add basic GUI

dsja612 committed:
Merge branch 'branch-Level-10'

dsja612 committed:
Fix bug where task status is not being saved

dsja612 committed:
Made user experience smoother

The GUI does not automatically close after the bye command is entered.
The size of the window is also too small to display the initial list of
supported commands.

The GUI will now automatically close after a 3 second delay. Also, the
GUI window is a bit bigger to accommodate long lists of text in the
future if needed.

dsja612 committed:
Add assertions for multiple files

dsja612 committed:
Improve on code quality

dsja612 committed:
Change Find command to use streams

dsja612 committed:
Merge pull request #1 from dsja612/branch-A-Assertions

Add assertions for multiple files

dsja612 committed:
Merge pull request #3 from dsja612/branch-A-Streams

Change Find command to use streams

dsja612 committed:
Merge branch 'master' into branch-A-CodeQuality

dsja612 committed:
Merge pull request #2 from dsja612/branch-A-CodeQuality

Improve on code quality

dsja612 committed:
Add command for user to find closest free day

dsja612 committed:
Merge pull request #4 from dsja612/branch-FindFreeTimes

Add command for user to find closest free day

dsja612 committed:
Update GUI to have improved aesthetics and resizing

dsja612 committed:
Add button image and made small tweaks to GUI

dsja612 committed:
Finalise and clean up GUI

dsja612 committed:
Update README and fixed minor bugs

dsja612 committed:
Add more UI improvements

dsja612 committed:
Add Ui.png

dsja612 committed:
Fix name of UI image

dsja612 committed:
Add JavaDocs to multiple files and improve code quality

dsja612 committed:
Fix some JavaDocs format issues

dsja612 committed:
Update README.md
